File: 卡码/代码随想录/1/04-字符串/C++/03-反转字符串中的单词/LeetCode_0151.cpp
//151. 反转字符串中的单词
#include <string>
using namespace std;

class Solution {
public:
  string reverseWords(string s) {
  	removeSpaces(s);
  	reverse(s,0,s.size()-1);
  	int start=0;
  	for(int i=0;i<=s.size();i++){	//每个单词反转
			if(i==s.size() || s[i]==' '){
				reverse(s,start,i-1);
				start=i+1;	//更新下一个起始点
			}
		}
		return s;
  }
  void reverse(string &s,int start,int end){
		for(int i=start,j=end;i<j;i++,j--)
			swap(s[i],s[j]);
	}
	void removeSpaces(string &s){
		int slow=0;
		for(int i=0;i<s.size();i++){
			if(s[i]!=' '){	//遇到空格就删除所有空格
				if(slow!=0) s[slow++]=' ';
				while(i<s.size()&&s[i]!=' ') s[slow++]=s[i++];
			}
		}
		s.resize(slow);
	}
};

File: 卡码/代码随想录/1/04-字符串/Java/src/_03_反转字符串中的单词/LeetCode_0151.java
//151. 反转字符串中的单词
package _03_反转字符串中的单词;

public class LeetCode_0151 {

}
class Solution {
  public String reverseWords(String s) {
  	StringBuilder builder = removeSpace(s);
  	reverse(builder,0,builder.length()-1);
  	int start=0,end=1,n=builder.length();
  	while(start<n) {
  		while(end<n&&builder.charAt(end)!=' ') end++;
  		reverse(builder,start,end-1);
  		start=end+1;
  		end=start+1;
  	}
  	return new String(builder);
  }
  public void reverse(StringBuilder builder,int start,int end) {
  	while(start<end) {
  		char tmp = builder.charAt(end);
  		builder.setCharAt(end,builder.charAt(start));
  		builder.setCharAt(start,tmp);
  		start++;
  		end--;
  	}
  }
  public StringBuilder removeSpace(String s) {
  	s = s.trim();
  	int start=0,end=s.length()-1;
  	StringBuilder builder = new StringBuilder();
  	while(start<=end) {
  		char c = s.charAt(start);
  		if(c!=' '||builder.charAt(builder.length()-1)!=' ') builder.append(c);
  		start++;
  	}
  	return builder;
  }
}